### PR TITLE
Testability

### PR DESCRIPTION
**Pull request by stonepreston**
Adds unit tests for the manager component of the juju autoscaler. This involved refactoring the code to be more testable via dependency injection and interfaces to facilitate mocking. This PR also adds a github workflow to run the tests, and a readme for the juju provider.

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/juju/juju_cloud_provider.go [hunk 1]
```diff
@@ -40,6 +41,15 @@ const (
...
 )
 
+// Note: struct fields must be public in order for unmarshal to
+// correctly populate the data.
+type jujuCloudConfig struct {
+	User      string   `yaml:"user"`
+	Password  string   `yaml:"password"`
+	Endpoints []string `yaml:"endpoints"`
+	CAcert    string   `yaml:"ca-cert"`
+}
```
Comment (addyess) [inline cluster-autoscaler/cloudprovider/juju/juju_cloud_provider.go:51]: awesome.  thank you for making them lowercase!

**Review by addyess — APPROVED** [review]
Let's be honest, i'm not a go expert.  But i can TOTALLY read this code and it's pretty clear what's going on.  Amazing improvement from my point of view

File: cluster-autoscaler/.github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+name: "Build"
```
Comment (Cynerva) [inline cluster-autoscaler/.github/workflows/test.yaml]: I think this file needs to be moved to the root of the repo, from

```
cluster-autoscaler/.github/workflows/test.yaml
```

to

```
.github/workflows/test.yaml
```

per GitHub documentation here: https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#about-workflow-triggers

> GitHub searches the `.github/workflows` directory in your repository for workflow files that are present in the associated commit SHA or Git ref of the event.

**Review by Cynerva — CHANGES_REQUESTED** [review]
Nice work. I just have one concern with the GitHub workflow file.

File: cluster-autoscaler/.github/workflows/test.yaml (hunk 2 above)
Comment (stonepreston) [reply cluster-autoscaler/.github/workflows/test.yaml]: Alright good catch. Ive moved the workflow up, I also disabled the other upstream workflows via the github UI. Unfortunately I was testing locally using go 1.17 and things were working fine. The go mod file specified 1.16 though. The tests were failing when running on GH using 1.16, so had to update go.mod to 1.17 which changed a bunch of vendored packages. Tests are passing now

**Review by Cynerva — APPROVED** [review]

**Pull request merged**